### PR TITLE
lib-user: Add Contributors CSV data export

### PR DESCRIPTION
**Pull request by mcbouslog**
## Package
- lib-user

## Describe your changes
- remove data export button from user stats
- refactor ContentLink, usePanoptesUser, and stats mock data for data export
- add Contributors data export functionality
- add convertStatsSecondsToHours utility function, refactor related components

## How to Review
_Helpful explanations that will make your reviewer happy:_
- What Zooniverse project should my reviewer use to review UX? n/a
- What user actions should my reviewer step through to review this PR?
  - lib-user: https://local.zooniverse.org:8080/?groups=1326962/contributors
  - app-root, staging: https://local.zooniverse.org:3000/groups/1326962/contributors
  - app-root, production: https://local.zooniverse.org:3000/groups/2620016/contributors?env=production
- Which storybook stories should be reviewed? n/a
- Are there plans for follow up PR’s to further fix this bug or develop this feature? yes, see linked GitHub project

# Checklist
_PR Creator - Please cater the checklist to fit the review needed for your code changes._
_PR Reviewer - Use the checklist during your review. Each point should be checkmarked or discussed before PR approval._

## General
- [ ] Tests are passing locally and on Github
- [ ] Documentation is up to date and changelog has been updated if appropriate
- [ ] You can `yarn panic && yarn bootstrap` or `docker-compose up --build` and FEM works as expected
- [ ] FEM works in all major desktop browsers: Firefox, Chrome, Edge, Safari (Use Browserstack account as needed)
- [ ] FEM works in a mobile browser

## General UX
Example Staging Project: [i-fancy-cats](https://local.zooniverse.org:3000/projects/brooke/i-fancy-cats)
- [ ] All pages of a FEM project load: Home Page, Classify Page, and About Pages
- [ ] Can submit a classification
- [ ] Can sign-in and sign-out
- [ ] The component is accessible
  - Can be used with a screen reader [BBC guide to VoiceOver](https://bbc.github.io/accessibility-news-and-you/assistive-technology/testing-steps/voiceover-mac.html)
  - Can be used from the keyboard [WebAIM guide to keyboard testing](https://webaim.org/techniques/keyboard/#testing)
  - It is passing accessibility checks in the storybook

## New Feature
- [ ] The PR creator has listed user actions to use when testing the new feature
- [ ] Unit tests are included for the new feature
- [ ] A storybook story has been created or updated
  - Example of SlideTutorial [component](https://github.com/zooniverse/front-end-monorepo/blob/master/packages/lib-classifier/src/components/Classifier/components/SlideTutorial/SlideTutorial.js) with docgen comments, and its [story](https://zooniverse.github.io/front-end-monorepo/@zooniverse/classifier/index.html?path=/docs/other-slidetutorial--default)

### CODE REVIEW

File: packages/lib-user/src/components/Contributors/helpers/getExportData.js [hunk 1]
```diff
@@ -0,0 +1,76 @@
...
+  projects,
+  users
+}) {
+  // create a list of all project names
+  let privateProjectIndex = 1
+  const projectNames = [...new Set(
+    stats.group_member_stats_breakdown.flatMap(member => {
+    return member.project_contributions.map(statsProject => {
+      const project = projects.find(project => project.id === statsProject.project_id.toString())
+      return project?.display_name || `Private Project ${privateProjectIndex++}`
```
Comment (mcbouslog) [inline packages/lib-user/src/components/Contributors/helpers/getExportData.js:14]: Currently, in zooniverse/front-end-monorepo/blob/lib-user_csv-export/packages/lib-user/src/components/Contributors/components/ContributorsList/ContributorsList.js#L51 it just notes "Private Project" for any and all private project display names, but should it include an incremented index there too? I lean towards yes, but not sure.

File: packages/lib-user/src/components/Contributors/helpers/generateExport.js [hunk 2]
```diff
@@ -0,0 +1,58 @@
...
+}) {
+  const data = getExportData({ projects, stats, users })
+
+  let str = ''
+
+  data.forEach((row) => {
+    str += row.map((col) => JSON.stringify(col)).join(',').concat('\n')
+  })
+
+  let newFilename = `${group.display_name}.data_export.${Date.now()}.csv`
```
Comment (mcbouslog) [inline packages/lib-user/src/components/Contributors/helpers/generateExport.js:23]: This needs to be updated with the `group.display_name` sanitized to remove emojis at minimum.

File: packages/lib-user/src/components/Contributors/Contributors.js [hunk 3]
```diff
@@ -0,0 +1,144 @@
...
+        <HeaderLink
+          href={`/groups/${group.id}`}
+          label='back'
+          primaryItem={true}
+        />
+      }
+    >
+      <ContentBox
+        linkLabel='Export all stats'
+        linkProps={{
```
Comment (mcbouslog) [inline packages/lib-user/src/components/Contributors/Contributors.js:101]: This should include `as: 'button'` and the ContentLink styling should be refactored accordingly.

File: packages/lib-user/src/components/Contributors/helpers/getExportData.js (hunk 1 above)
Comment (mcbouslog) [reply packages/lib-user/src/components/Contributors/helpers/getExportData.js:14]: Addressed in e0c38ab .

File: packages/lib-user/src/components/Contributors/helpers/generateExport.js (hunk 2 above)
Comment (mcbouslog) [reply packages/lib-user/src/components/Contributors/helpers/generateExport.js]: Refactored with 782f3f5

**Review by goplayoutside3 — COMMENTED** [review]
A couple of things to double check:

1. In the `getExportData` function, are you exporting time as hours? See my comment in that file where there's a mismatch of units.

2. I used group 1326962 (staging) as an example, and it's a private, aggregate data group. I noticed that when using admin mode, the Top Contributors section appears (which I think is expected) but the link to 'See all contributors and detailed stats' is not formed correctly. The href is missing the group number. I think this is a bug that might affect all group admins, as well as Zooniverse admins?

File: packages/lib-user/src/components/Contributors/helpers/getExportData.js [hunk 4]
```diff
@@ -0,0 +1,76 @@
...
+
+    // iterate over each project and add the member stats for that project or 0 if the member has no stats for that project
+    for (const projectName of projectNames) {
+      const projectStats = member.project_contributions
+        .find(statsProject => {
+          const project = projects.find(project => project.id === statsProject.project_id.toString())
+          return project?.display_name === projectName
+        })
+      row.push(projectStats?.count || 0)
+      row.push(projectStats?.session_time || 0)
```
[the remark this answers is not in the document]
Comment (mcbouslog) [reply packages/lib-user/src/components/Contributors/helpers/getExportData.js]: Yes, that's correct! Fixed in 5f80341

**Comment by mcbouslog** [comment]
1. I added a utility function as previously discussed to convert stats seconds to hours, called `convertStatsSecondsToHours` and refactored `getExportData` as well as other related components accordingly
2. The TopContributors component expected the groupId as a prop. I added it back, I might have removed it in a rebase in error, but should now work as expected

File: packages/lib-user/src/components/Contributors/helpers/generateExport.js [hunk 5]
```diff
@@ -0,0 +1,61 @@
...
+  let str = ''
+
+  data.forEach((row) => {
+    str += row.map((col) => JSON.stringify(col)).join(',').concat('\n')
+  })
+
+  // The following regexp sanitizes the group name by removing all non-alphanumeric characters (i.e. emojis, spaces, punctuation, etc.)
+  let sanitizedGroupName = group.display_name.replace(/[^a-zA-Z0-9]/g, '')
+
+  let newFilename = `${sanitizedGroupName}.data_export.${Date.now()}.csv`
```
Comment (goplayoutside3) [inline packages/lib-user/src/components/Contributors/helpers/generateExport.js:26]: `Date.now()` works, and for example I successfully exported data with the file name `TestGroup5.data_export.1718129145512.csv`. Just to double check though, `Date.now()` is not really human-readable. Was the intent to insert a date into the filename to simply make it unique? Or to use a human-readable date like YYYY_MM_DD? For example, lib-content now uses `dayjs` for easy conversion in CommunityContainer if you're looking for a lightweight third-party package.

**Review by goplayoutside3 — APPROVED** [review]
Looks good! Just one thing to double check with the file name, but otherwise 👍

File: packages/lib-user/src/components/Contributors/helpers/generateExport.js (hunk 5 above)
Comment (mcbouslog) [reply packages/lib-user/src/components/Contributors/helpers/generateExport.js:26]: Intent was just to make filename unique, but I think the date would be helpful, especially if easy to do with `dayjs`. I'll confirm including date is preferable with Sean/Laura and update accordingly 👍 .

**Pull request merged**